sharding/proposer/service.go

```diff
@@ -46,42 +60,57 @@ func (p *Proposer) Start() {
 // Stop the main loop for proposing collations.
 func (p *Proposer) Stop() error {
 	log.Info(fmt.Sprintf("Stopping proposer service in shard %d", p.shardID))
+	defer p.cancel()
+	p.txpoolSub.Unsubscribe()
 	return nil
 }
 
+// proposeCollations listens to the transaction feed and submits collations over an interval.
 func (p *Proposer) proposeCollations() {
+	requests := make(chan *types.Transaction)
+	p.txpoolSub = p.txpool.TransactionsFeed().Subscribe(requests)
+
+	for {
+		ctx, cancel := context.WithTimeout(p.ctx, 10*time.Second)
+		select {
+		case tx := <-requests:
+			log.Info(fmt.Sprintf("Received transaction: %x", tx.Hash()))
+			if err := p.createCollation(ctx, []*types.Transaction{tx}); err != nil {
+				log.Error(fmt.Sprintf("Create collation failed: %v", err))
+			}
+		case <-ctx.Done():
+			log.Info("Context canceled")
+		case err := <-p.txpoolSub.Err():
+			log.Error(fmt.Sprintf("Subscriber failed: %v", err))
+		}
 
-	// TODO: Receive TXs from shard TX generator or TXpool (Github Issues 153 and 161)
-	var txs []*types.Transaction
-	for i := 0; i < 10; i++ {
-		data := make([]byte, 1024)
-		rand.Read(data)
-		txs = append(txs, types.NewTransaction(0, common.HexToAddress("0x0"),
-			nil, 0, nil, data))
+		cancel()
+		return
```

Comment:
Eliminate these two lines. Cancel will be called only if the service stops in the `Stop()` function.